File: Graph Theory/leetcode_322.py

# Problem link ===>> https://leetcode.com/problems/reconstruct-itinerary/description/?envType=problem-list-v2&envId=21rco0xh
# submission link == =>> https://leetcode.com/problems/reconstruct-itinerary/submissions/1725752461/?envType=problem-list-v2&envId=21rco0xh

import sys
import math
import bisect
import heapq
from collections import defaultdict, deque, OrderedDict, Counter
from functools import lru_cache, reduce
from itertools import permutations, combinations, accumulate
from typing import List, Tuple, Dict, Set, Optional, Union, Any

MOD = 10**9 + 7
INF = float('inf')
input = sys.stdin.read
print = sys.stdout.write


class Solution:
    def __init__(self):
        self.flightGraph = defaultdict(list)
        self.itinerary = []
    
    def dfs(self, airport: str)-> None:
        while self.flightGraph[airport]:
            next_airport = self.flightGraph[airport].pop()
            self.dfs(next_airport)
        self.itinerary.append(airport)
    
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        for src, dest in tickets:
            self.flightGraph[src].append(dest)
        
        for src in self.flightGraph:
            self.flightGraph[src].sort(reverse=True)
        
        self.dfs("JFK")
        return self.itinerary[::-1]

def main():
    t = int(input())
    for _ in range(t):
        n = int(input())

if __name__ == "__main__":
    solution = Solution()

File: Graph Theory/leetcode_322.cpp

// Problem link ===>> https://leetcode.com/problems/reconstruct-itinerary/description/?envType=problem-list-v2&envId=21rco0xh

#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;
#define ll long long
#define ld long double
#define endl '\n'
const int MOD = 1e9 + 7;
const double EPS = 1e-9;
const double PI = acos(-1.0);
template <typename T>
using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
template <typename K, typename V>
using ordered_map = tree<
    K, V, less<K>, rb_tree_tag, tree_order_statistics_node_update>;

class Solution
{
private:
    unordered_map<string, multiset<string, greater<string>>> flightGraph;
    vector<string> itinerary;

public:
    void dfs(string airport)
    {
        auto &destinations = flightGraph[airport];

        while (!destinations.empty())
        {
            auto nextDestination = *(--destinations.end());
            destinations.erase(--destinations.end());
            dfs(nextDestination);
        }
        itinerary.push_back(airport);
    }

    vector<string> findItinerary(vector<vector<string>> &tickets)
    {
        for (int i = 0; i < tickets.size(); i++)
        {
            string from = tickets[i][0];
            string to = tickets[i][1];
            flightGraph[from].insert(to);
        }

        dfs("JFK");

        reverse(itinerary.begin(), itinerary.end());
        return itinerary;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);

    ll t;
    cin >> t;
    while (t--)
    {
        ll n;
        cin >> n;
    }
    return 0;
}